cocotb/log.py

```diff
@@ -65,89 +70,43 @@ def __init__(self, name):
-                                            level,
-                                            info.filename,
-                                            info.lineno,
-                                            msg,
-                                            args,
-                                            None,
-                                            info.function,
-                                            extra)
-            self.logger.handle(record)
-
-    def _willLog(self, level):
-        """ This is for user from the C world
-            it allows a check on if the message will
-            be printed. Saves doing lots of work
-            for no reason.
+    def _log_from_c(self, level, filename, lineno, msg, function):
         """
-        return self.logger.isEnabledFor(level)
-
-    def _printRecord(self, level, filename, lineno, msg, function):
-        """ This is for use from the C world and will
-            be printed regardless
+        This is for use from the C world, and allows us to insert C stack
+        information.
         """
-        if self.logger.isEnabledFor(level):
-            record = self.logger.makeRecord(self._log_name,
-                                            level,
-                                            filename,
-                                            lineno,
-                                            msg,
-                                            None,
-                                            None,
-                                            function)
-            self.logger.handle(record)
-
-    def warn(self, msg, *args, **kwargs):
-        self._makeRecord(logging.WARNING, msg, args, **kwargs)
-
-    def warning(self, msg, *args, **kwargs):
-        self._makeRecord(logging.WARNING, msg, args, **kwargs)
-
-    def debug(self, msg, *args, **kwargs):
-        self._makeRecord(logging.DEBUG, msg, args, **kwargs)
-
-    def error(self, msg, *args, **kwargs):
-        self._makeRecord(logging.ERROR, msg, args, **kwargs)
-
-    def critical(self, msg, *args, **kwargs):
-        self._makeRecord(logging.CRITICAL, msg, args, **kwargs)
-
-    def info(self, msg, *args, **kwargs):
-        self._makeRecord(logging.INFO, msg, args, **kwargs)
-
-    def __getattr__(self, attribute):
-        """Forward any other attribute accesses on to our logger object"""
-        return getattr(self.logger, attribute)
+        if self.isEnabledFor(level):
+            record = self.makeRecord(
+                self.name,
+                level,
+                filename,
+                lineno,
+                msg,
+                None,
+                None,
+                function
+            )
+            self.handle(record)
+
+    @property
+    def logger(self):
+        warnings.warn(
+            "the .logger attribute, now that `SimLog` returns a native logger "
+            "directly.",
+            DeprecationWarning)
```

Comment:
Our testsuite had some code of the form `x._log.logger.setLevel(...)`.

This was never the right way to spell it, and `x._log.setLevel(...)` works before and after this change.

However, `.logger` does not start with an `_`, so in principle someone could be using it. Adding this property with a deprecation warning ensures this change does not break existing incorrect code.